oracle/reactor.go

```diff
@@ -0,0 +1,246 @@
+			logrus.Errorf("failed signature verification for validator: %v, skipping gossip", val.Address)
+			return
+		}
+
+		preLockTime := time.Now().UnixMilli()
+		oracleR.OracleInfo.GossipVoteBuffer.UpdateMtx.Lock()
+		currentGossipVote, ok := oracleR.OracleInfo.GossipVoteBuffer.Buffer[pubKey.Address().String()]
+
+		if !ok {
+			// first gossipVote entry from this validator
+			oracleR.OracleInfo.GossipVoteBuffer.Buffer[pubKey.Address().String()] = msg
+		} else {
+			// existing gossipVote entry from this validator
+			previousTimestamp := currentGossipVote.SignedTimestamp
+			newTimestamp := msg.SignedTimestamp
+			// only replace if the gossipVote received has a later timestamp than our current one
+			if newTimestamp > previousTimestamp {
+				oracleR.OracleInfo.GossipVoteBuffer.Buffer[pubKey.Address().String()] = msg
+			}
+		}
+		oracleR.OracleInfo.GossipVoteBuffer.UpdateMtx.Unlock()
+		postLockTime := time.Now().UnixMilli()
+		diff := postLockTime - preLockTime
+		if diff > 100 {
+			logrus.Warnf("WARNING!!! Receiving gossip lock took %v milliseconds", diff)
+		}
+	default:
+		logrus.Warn("unknown message type", "src", e.Src, "chId", e.ChannelID, "msg", e.Message)
+		oracleR.Switch.StopPeerForError(e.Src, fmt.Errorf("oracle cannot handle message of type: %T", e.Message))
+		return
+	}
+
+	// broadcasting happens from go routines per peer
+}
+
+// PeerState describes the state of a peer.
+type PeerState interface {
+	GetHeight() int64
+}
+
+// // Send new oracle votes to peer.
+func (oracleR *Reactor) broadcastVoteRoutine(peer p2p.Peer) {
+	// gossip votes every x milliseconds, where x = Config.GossipInterval
+	interval := oracleR.OracleInfo.Config.GossipInterval
+
+	for {
+		// In case of both next.NextWaitChan() and peer.Quit() are variable at the same time
+		if !oracleR.IsRunning() || !peer.IsRunning() {
+			return
+		}
+		select {
+		case <-peer.Quit():
+			return
+		case <-oracleR.Quit():
+			return
+		default:
+		}
+
+		// Make sure the peer is up to date.
+		_, ok := peer.Get(types.PeerStateKey).(PeerState)
+		if !ok {
+			// Peer does not have a state yet. We set it in the consensus reactor, but
+			// when we add peer in Switch, the order we call reactors#AddPeer is
+			// different every time due to us using a map. Sometimes other reactors
+			// will be initialized before the consensus reactor. We should wait a few
+			// milliseconds and retry.
+			time.Sleep(PeerCatchupSleepIntervalMS * time.Millisecond)
+			continue
+		}
+
+		// only gossip votes that are younger than the latestAllowableTimestamp, which is the max(earliest block timestamp collected, current time - maxOracleGossipAge)
+		latestAllowableTimestamp := time.Now().Unix() - int64(oracleR.OracleInfo.Config.MaxOracleGossipAge)
+		if len(oracleR.OracleInfo.BlockTimestamps) > 0 && oracleR.OracleInfo.BlockTimestamps[0] > latestAllowableTimestamp {
+			latestAllowableTimestamp = oracleR.OracleInfo.BlockTimestamps[0]
```

Comment:
need to check that buffer is full before replacing

Reply:
resolved in https://github.com/Switcheo/cometbft/pull/1/commits/3babce32ddcfbbf971a0da6717e8f2d1aec5529a